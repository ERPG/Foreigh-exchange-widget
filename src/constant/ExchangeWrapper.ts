export const defaultCurrecySelection = {
  baseRate: 1,
  currentRate: 0,
};

export const INITIAL_AMOUNT: number = 0;
