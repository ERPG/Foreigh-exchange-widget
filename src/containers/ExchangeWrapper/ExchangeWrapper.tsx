import { Dispatch, useCallback, useEffect, useState } from 'react';
import { SelectChangeEvent } from '@mui/material';

import CurrencyBox from '../../components/CurrencyBox/CurrencyBox';
import Button from '../../common/components/Button/Button';

import { useDispatch, useSelector } from 'react-redux';
import { getForeighRates } from '../../store/ForeighRates/actions/getForeighRates';
import { selectForeighRates } from '../../store/ForeighRates/reducer';
import {
  Currencies,
  CurrencyRate,
  ForeighRates,
} from '../../models/ForeighRates';
import {
  defaultCurrecySelection,
  INITIAL_AMOUNT,
} from '../../constant/ExchangeWrapper';

export type CurrencySelection = {
  baseRate: number;
  currentRate: number;
};

const ExchangeWrapper = () => {
  const dispatch: Dispatch<any> = useDispatch();
  const foreighRates: ForeighRates | undefined =
    useSelector(selectForeighRates);

  useEffect(() => {
    dispatch(getForeighRates());
  }, [dispatch]);

  const handleCurrencySwitch = useCallback(
    (baseR: number, currentR: number) => {
      const newBaseFXRate: CurrencyRate<number> | undefined =
        getFXCurrencyByRate(baseR);
      const newCurrentFXRate: CurrencyRate<number> | undefined =
        getFXCurrencyByRate(currentR);

      SetCurrencies({
        base: newBaseFXRate?.currency,
        current: newCurrentFXRate?.currency,
      });
      dispatch(getForeighRates(newBaseFXRate?.currency));
    },
    [dispatch, foreighRates]
  );

  const [selectValue, SetSelectValue] = useState<CurrencySelection>(
    defaultCurrecySelection
  );

  const handleSelectState = (state: CurrencySelection): void => {
    const { baseRate, currentRate } = state;
    SetSelectValue({
      baseRate,
      currentRate,
    });
  };

  const handleSelectChange = (event: SelectChangeEvent<string>): void => {
    const rateSelected = Number(event.target.value);
    const { baseRate, currentRate } = selectValue;
    const isCurrentEqualToBase = rateSelected === baseRate;
    const isBaseEqualToCurrent = rateSelected === currentRate;
    if (isCurrentEqualToBase) {
      handleCurrencySwitch(currentRate, rateSelected);
    } else if (isBaseEqualToCurrent) {
      handleCurrencySwitch(rateSelected, baseRate);
    } else {
      const selectName = event.target.name;
      const selectRate = selectName.substring(0, selectName.indexOf('-'));
      const { baseAmount } = amount;
      handleSelectState({ ...selectValue, [selectRate]: rateSelected });
      calculateCurrentAmount(baseAmount, rateSelected);
    }
  };

  const [currencies, SetCurrencies] = useState<{
    base: Currencies | undefined;
    current: Currencies | undefined;
  }>({ base: Currencies.EUR, current: Currencies.USD });

  useEffect(() => {
    if (foreighRates) {
      const { base, current } = currencies;
      const newBaseFXRate: CurrencyRate<number> | undefined =
        getFXRateByCurrency(base);
      const newCurrentFXRate: CurrencyRate<number> | undefined =
        getFXRateByCurrency(current);
      const areRatesValid = newBaseFXRate?.rate === 1 && newCurrentFXRate?.rate;

      if (areRatesValid) {
        handleSelectState({
          baseRate: newBaseFXRate?.rate,
          currentRate: newCurrentFXRate?.rate,
        });
        handleAmountState({
          baseAmount: INITIAL_AMOUNT,
          resultAmount: INITIAL_AMOUNT,
        });
      }
    }
  }, [foreighRates]);

  const getFXCurrencyByRate = (
    rate: number | undefined
  ): CurrencyRate<number> | undefined => {
    return foreighRates
      ?.getExchangeRatesArr()
      .find((elem: CurrencyRate<number>) => elem.rate === rate);
  };

  const getFXRateByCurrency = (
    currency: Currencies | undefined
  ): CurrencyRate<number> | undefined => {
    return foreighRates
      ?.getExchangeRatesArr()
      .find((elem: CurrencyRate<number>) => elem.currency === currency);
  };

  const [amount, handleAmount] = useState<{
    baseAmount: number;
    resultAmount: number;
  }>({ baseAmount: INITIAL_AMOUNT, resultAmount: INITIAL_AMOUNT });

  const handleAmountState = (state: {
    baseAmount: number;
    resultAmount: number;
  }): void => {
    const { baseAmount, resultAmount } = state;
    handleAmount({
      baseAmount,
      resultAmount,
    });
  };

  const calculateCurrentAmount = (amount: number, rate: number): void => {
    const result: number = amount * rate;

    handleAmountState({
      baseAmount: amount,
      resultAmount: result.toFixed(2) as unknown as number,
    });
  };

  const calculateBaseAmount = (amount: number, rate: number): void => {
    const result: number = amount / rate;

    handleAmountState({
      baseAmount: amount,
      resultAmount: result.toFixed(2) as unknown as number,
    });
  };

  const handleInputBaseChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ): void => {
    const inputBaseValue: number = Number(
      event.target.value.replace(/[^0-9]/g, '')
    );
    calculateCurrentAmount(inputBaseValue, selectValue.currentRate);
  };

  const handleInputResultChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ): void => {
    const inputResultValue: number = Number(
      event.target.value.replace(/[^0-9]/g, '')
    );
    calculateBaseAmount(inputResultValue, selectValue.currentRate);
  };

  const handleForeighExchangeAmount = () => {};

  return (
    <div style={{ marginTop: 80 }}>
      <CurrencyBox
        id="baseRate"
        currencyRates={foreighRates?.getExchangeRatesArr()}
        amount={amount.baseAmount}
        handleAmount={handleInputBaseChange}
        selectValue={selectValue.baseRate}
        handleSelectChange={handleSelectChange}
        inputLabel={'Sending'}
      />
      <CurrencyBox
        id="currentRate"
        currencyRates={foreighRates?.getExchangeRatesArr()}
        amount={amount.resultAmount}
        handleAmount={handleInputResultChange}
        selectValue={selectValue.currentRate}
        handleSelectChange={handleSelectChange}
        inputLabel={'Receiving'}
      />
      <div style={{ marginTop: 80, padding: 16 }}>
        <Button name="exchange-button" onClick={handleForeighExchangeAmount}>
          Send
        </Button>
      </div>
    </div>
  );
};

export default ExchangeWrapper;
