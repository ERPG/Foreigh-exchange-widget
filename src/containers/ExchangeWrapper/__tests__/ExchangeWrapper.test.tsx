import * as Redux from 'react-redux';
import { generateAppStateMock } from '../../../../__tests__/__fixtures__/generateAppStateMock';
import { renderStoreHelper } from '../../../../__tests__/__utils__/renderStoreHelper';

import ExchangeWrapper from '../ExchangeWrapper';

describe('ExchangeWrapper component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('Should render component without errors', () => {
    const { container } = renderStoreHelper({
      component: <ExchangeWrapper />,
      rootState: generateAppStateMock(),
    });

    expect(container).toMatchSnapshot();
  });

  it('Should call redux hooks on init', () => {
    const useSelectorSpy = jest.spyOn(Redux, 'useSelector');
    const useDispatchSpy = jest.spyOn(Redux, 'useDispatch');

    const { container } = renderStoreHelper({
      component: <ExchangeWrapper />,
      rootState: generateAppStateMock(),
    });

    expect(container).toBeDefined();
    expect(useSelectorSpy).toBeCalled();
    expect(useDispatchSpy).toBeCalled();
  });
});
