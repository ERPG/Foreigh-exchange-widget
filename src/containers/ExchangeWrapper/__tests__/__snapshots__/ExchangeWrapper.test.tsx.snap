// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeWrapper component Should render component without errors 1`] = `
<div>
  <div
    style="margin-top: 80px;"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
      style="text-align: center; height: 80px; line-height: 60px; margin: 16px; display: flex; justify-content: space-around; align-items: center;"
    >
      <div>
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
            data-shrink="true"
            for="currency-amount"
            id="currency-amount-label"
          >
            Sending
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
              data-fx="fx-amount-input"
              id="currency-amount"
              inputmode="numeric"
              name="baseRate-input"
              type="text"
              value="0"
            />
          </div>
        </div>
      </div>
      <div
        style="margin-top: 16px;"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-1v3hyuc-MuiFormControl-root"
        >
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl makeStyles-root-1 MuiSelect-root css-1480iag-MuiInputBase-root-MuiInput-root"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-label="Without label"
              aria-labelledby="mui-component-select-baseRate-select"
              class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
              id="mui-component-select-baseRate-select"
              role="button"
              tabindex="0"
            >
              EUR
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              data-fx="fx-currency-selection"
              name="baseRate-select"
              tabindex="-1"
              value="1"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
      style="text-align: center; height: 80px; line-height: 60px; margin: 16px; display: flex; justify-content: space-around; align-items: center;"
    >
      <div>
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiFormLabel-filled"
            data-shrink="true"
            for="currency-amount"
            id="currency-amount-label"
          >
            Receiving
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
              data-fx="fx-amount-input"
              id="currency-amount"
              inputmode="numeric"
              name="currentRate-input"
              type="text"
              value="0"
            />
          </div>
        </div>
      </div>
      <div
        style="margin-top: 16px;"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-1v3hyuc-MuiFormControl-root"
        >
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl makeStyles-root-1 MuiSelect-root css-1480iag-MuiInputBase-root-MuiInput-root"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-label="Without label"
              aria-labelledby="mui-component-select-currentRate-select"
              class="MuiSelect-select MuiSelect-standard MuiInput-input MuiInputBase-input css-1rxz5jq-MuiSelect-select-MuiInputBase-input-MuiInput-input"
              id="mui-component-select-currentRate-select"
              role="button"
              tabindex="0"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              data-fx="fx-currency-selection"
              name="currentRate-select"
              tabindex="-1"
              value="0"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      style="margin-top: 80px; padding: 16px;"
    >
      <button
        class="makeStyles-root-2 MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-disableElevation MuiButton-fullWidth MuiButtonBase-root css-1o1g6k0-MuiButtonBase-root-MuiButton-root"
        data-fx="fx-button"
        name="exchange-button"
        tabindex="0"
        type="button"
      >
        <span>
          Send
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
